checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -43,104 +82,596 @@ public class MustCallInferenceLogic {
+   *     the node does not
+   */
+  private Set<ResourceAlias> getResourceAliasOfArgument(Set<Obligation> obligations, Node arg) {
+    Node tempVar = mcca.getTempVarOrNode(arg);
+    if (!(tempVar instanceof LocalVariableNode)) {
+      return Collections.emptySet();
+    }
+
+    Obligation argumentObligation =
+        MustCallConsistencyAnalyzer.getObligationForVar(obligations, (LocalVariableNode) tempVar);
+    if (argumentObligation == null) {
+      return Collections.emptySet();
+    }
+    return argumentObligation.resourceAliases;
+  }
+
+  /**
+   * If the receiver of {@code mNode} is a candidate owning field and the method invocation
+   * satisfies the field's must-call obligation, then adds that field to the {@link
+   * #owningFieldToECM} set.
+   *
+   * @param obligations the set of obligations to search in
+   * @param mNode the MethodInvocationNode
+   */
+  private void checkMethodInvocation(Set<Obligation> obligations, MethodInvocationNode mNode) {
+    if (enMethodElt == null) {
+      return;
+    }
+
+    List<? extends VariableTree> paramsOfEnclosingMethod = enMethodTree.getParameters();
+
+    checkArgsOfMethodCall(obligations, mNode, paramsOfEnclosingMethod);
+    checkReceiverOfMethodCall(obligations, mNode, paramsOfEnclosingMethod);
+    checkIndirectCalls(obligations, mNode, paramsOfEnclosingMethod);
+  }
+
+  /**
+   * Checks if a MustCall obligation of the element is satisfied via the given method call. A
+   * MustCall obligation of an element is satisfied if the called-methods set contains the target of
+   * its must-call obligation.
+   *
+   * @param mNode The method invocation node being checked for satisfaction of the MustCall
+   *     obligation.
+   * @param varElt The element representing the variable annotated with the MustCall annotation.
+   * @param target The target of the MustCall obligation, represented as a JavaExpression.
+   * @return {@code true} if the MustCall obligation is satisfied, {@code false} otherwise.
+   */
+  private boolean mustCallObligationSatisfied(
+      MethodInvocationNode mNode, Element varElt, JavaExpression target) {
+
+    List<String> mustCallValues = typeFactory.getMustCallValue(varElt);
+    if (mustCallValues.size() != 1) {
       // TODO: generalize this to MustCall annotations with more than one element.
-      if (mustCallValues.size() == 1
-          && mustCallValues.contains(method.getSimpleName().toString())) {
-        owningFields.add((VariableElement) receiverEl);
+      return false;
+    }
+
+    AccumulationStore cmStoreAfter = typeFactory.getStoreAfter(mNode);
+    @Nullable AccumulationValue cmValue =
+        cmStoreAfter == null ? null : cmStoreAfter.getValue(target);
+    AnnotationMirror cmAnno = null;
+    if (cmValue != null) {
+      for (AnnotationMirror anno : cmValue.getAnnotations()) {
+        if (AnnotationUtils.areSameByName(
+            anno, "org.checkerframework.checker.calledmethods.qual.CalledMethods")) {
```

Comment:
Is there a reason to restrict this to only `@CalledMethods` annotations? You're going to use the CMC's subtyping check anyway, which can reason about other annotations in the checker (like `@CalledMethodsPredicate`).

Also, do you want to use the facts carried in `AccumulationValue` here? I think you probably do, if you want to support type variables. You should take a look at the change I made in the consistency checker in #5943 and consider making a similar change here.

Reply:
I made it consistent with your changes in consistency checker.